zilencer/views.py

```diff
@@ -672,6 +676,71 @@ def update_remote_realm_data_for_server(
     RemoteRealmAuditLog.objects.bulk_create(remote_realm_audit_logs)
 
 
+def get_billable_realm_uuids(realms: List[RealmDataForAnalytics]) -> List[UUID]:
+    billable_realm_uuids = []
+    for realm in realms:
+        # TODO: Remove the `zulipinternal` string_id check once no server is on 8.0-beta.
+        if (
+            realm.is_system_bot_realm
+            or realm.host.startswith("zulipinternal.")
+            or realm.host.startswith("analytics.")
+        ):
+            continue
+        billable_realm_uuids.append(realm.uuid)
+
+    return billable_realm_uuids
+
+
+@transaction.atomic
+def handle_customer_migration_from_server_to_realms(
+    server: RemoteZulipServer, realms: List[RealmDataForAnalytics]
+) -> None:
+    server_billing_session = RemoteServerBillingSession(server)
+    server_customer = server_billing_session.get_customer()
+    if server_customer is None:
+        return
+    server_plan = get_current_plan_by_customer(server_customer)
+    realm_uuids = get_billable_realm_uuids(realms)
+    if not realm_uuids:
+        return
+
+    if (
+        server_plan is not None
+        and server_plan.tier == CustomerPlan.TIER_SELF_HOSTED_LEGACY
+        and server_plan.status == CustomerPlan.ACTIVE
+    ):
+        assert server.plan_type == RemoteZulipServer.PLAN_TYPE_SELF_HOSTED
+        assert server_plan.end_date is not None
+        remote_realms = RemoteRealm.objects.filter(
+            uuid__in=realm_uuids, server=server, plan_type=RemoteRealm.PLAN_TYPE_SELF_HOSTED
+        )
+
+        # Verify that all the realms are on self hosted plan.
+        assert remote_realms.count() == len(realm_uuids)
+
+        # End existing plan for server.
+        server_plan.status = CustomerPlan.ENDED
+        server_plan.save(update_fields=["status"])
+
+        # Create new legacy plan for each remote realm.
+        for remote_realm in remote_realms:
+            RemoteRealmBillingSession(remote_realm).migrate_customer_to_legacy_plan(
+                server_plan.billing_cycle_anchor, server_plan.end_date
+            )
+    else:
+        # We only do this migration if there is only one realm besides the system bot realm.
+        assert len(realm_uuids) == 1
+        remote_realm = RemoteRealm.objects.get(
+            uuid=realm_uuids[0], plan_type=RemoteRealm.PLAN_TYPE_SELF_HOSTED
+        )
+        # Migrate customer from server to remote realm if there is only one realm.
+        server_customer.remote_realm = remote_realm
+        server_customer.remote_server = None
+        server_customer.save(update_fields=["remote_realm", "remote_server"])
+        server.plan_type = RemoteZulipServer.PLAN_TYPE_SELF_HOSTED
+        server.save(update_fields=["plan_type"])
```

Comment:
Do we need to change `remote_realm.plan_type = server.plan_type` before changing `server.plan_type`? Otherwise we might be left with the plan being weird.

Reply:
Yeah, nice catch.